### PR TITLE
Avoid hanging forever after a parallel job was killed

### PR DESCRIPTION
**Pull request by daniel-wer**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Replace `multiprocessing.pool` with [concurrent.futures.ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor) to avoid deadlocks.

In a `multiprocessing.pool`, if a process terminates in a non-clean fashion (for example, due to OOM or a segmentation fault), the pool will silently replace said process, but the work that the process was supposed to do will never be done, causing pylint to hang indefinitely. The `concurrent.futures.ProcessPoolExecutor` will raise a [BrokenProcessPool exception](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.process.BrokenProcessPool) in that case, avoiding the hang.

See https://github.com/PyCQA/pylint/issues/3899#issuecomment-732712793 for reproduction instructions.

The hangs caused by this issue ate up quite a few CI minutes for us over time. Although this doesn't fix the underlying sporadic issue (which is likely due to OOM) at least pylint will fail early instead of hanging forever.

A followup could be to recover from this type of error by catching the `BrokenProcessPool` exception and then instantiating a new `ProcessPoolExecutor` to continue the linting.

Closes https://github.com/PyCQA/pylint/issues/3899

### CODE REVIEW

**Comment by Pierre-Sassoulas** [comment]
This look promising, thank you for opening the PR !

**Comment by daniel-wer** [comment]
Thank _you_ for your work maintaining this great tool :pray: The tests were failing due to a silly mistake, should be fixed.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for fixing the tests :) Would you mind adding a changelog's fragment please ? Also I don't know if it's feasible but could we imagine an automated test where we kill a job to test the original issue?

**Comment by Pierre-Sassoulas** [comment]
> However, I'm not yet sure how to design the test where we kill a job to test the original issue. 

I don't know either. We're going to hit timing /concurrency issues and flakiness if we do it with an external command to kill a process launched inside our test. I never did that with pytest I was hoping a magical pytest fixture would exists to be honest 😄Maybe a manual test is enough ?

File: pylint/lint/parallel.py
Comment (DanielNoord) [inline pylint/lint/parallel.py:171]: Why did you remove the close and join here? These were added explicitly at some point, even though the context manager should handle these normally. Is there a specific reason for removing them? Doesn't `ProccessPoolExecutor` support them?

**Review by DanielNoord — CHANGES_REQUESTED** [review]

File: pylint/lint/parallel.py
Comment (daniel-wer) [reply pylint/lint/parallel.py:171]: > Is there a specific reason for removing them? Doesn't ProccessPoolExecutor support them?

Exactly, these are not supported by `ProccessPoolExecutor`. The context manager calls [shutdown](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.shutdown) upon exiting which is explicitly mentioned in the docs.

**Comment by DanielNoord** [comment]
@daniel-wer Do you understand the failing test?

**Comment by daniel-wer** [comment]
@DanielNoord I'm not entirely sure. My first suspicion was that it is timing-related. Since I configured the test to timeout after 1 second, it could happen that the time until the first pool process is initialized (and the `TypeError: 'int' object is not iterable` is thrown which leads to a `BrokenProcessPool` exception) is more than 1 second.

However, the captured stderr indicates that the `TypeError` was already raised. In all other configurations this led to the expected `BrokenProcessPool` exception. Might be that the timing was very unlucky in this case :thinking: 

We could try to configure a more generous timeout of 5 seconds since the timeout will only be triggered if the test fails and won't affect normal test run times. This way we could rule out timing-related issues. What do you think?

**Comment by DanielNoord** [comment]
Let's try the higher timeout indeed! Sorry for the last response!

**Review by DanielNoord — COMMENTED** [review]
I just did a check for references to `multiprocessing` and saw that we have a couple of checks that test whether it is available. Is this also a concern for `concurrency`?

Edit: It seems like it is, so we should probably change all those references and tests as well.

**Comment by daniel-wer** [comment]
@DanielNoord Thank you, I was not aware of that. Fixed.

**Comment by daniel-wer** [comment]
Could you give me a pointer to the relevant files? I had a look and from what I can see the `multiprocessing` module is used in some of the functional checker tests or to call `.current_process()` or `.cpu_count()`. I think neither of these usages can be replaced with the `concurrent` module (which doesn't have the `.current_process()` or `.cpu_count()` functionality).

The only thing I can see is that in `pylint/lint/run.py` in addition to `multiprocessing.cpu_count()` the existence of the `concurrent` module should be checked to fallback to a single process if it is not available, but I don't see where `multiprocessing` imports could be removed.

**Comment by daniel-wer** [comment]
@DanielNoord 

> The last thing was indeed the only thing I thinking of.

Ok great, I adapted the check.

> Just to double check: does the determining of cpus with multiprocessing also work for the amount of workers that concurrent can use?

The [docs](https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor) state that the "ProcessPoolExecutor uses the multiprocessing module", so yes that should be the case. Also, this kind of functionality was not duplicated for the `concurrent.futures` module which is rather an alternative interface for interacting with a process pool but not changing the underlying process fundamentals like `current_process()` or `cpu_count()`.

@Pierre-Sassoulas

> Maybe a manual test is enough ?

Using the latest version of this PR, I again confirmed that killing a random pylint worker process during a parallel linting no longer leads to a deadlock, but immediately exits with a `BrokenProcessPool` exception, using the reproduction instructions from https://github.com/PyCQA/pylint/issues/3899#issuecomment-732712793.

**Comment by DanielNoord** [comment]
I hope to find time to review this somewhere this weekend!

File: tests/test_check_parallel.py
Comment (DanielNoord) [inline tests/test_check_parallel.py:35]:
```suggestion
from concurrent.futures import ProcessPoolExecutor
from concurrent.futures.process import BrokenProcessPool
```

Let's make this necessary for tests still.

File: pylint/lint/parallel.py
Comment (DanielNoord) [inline pylint/lint/parallel.py:24]: This is one use of this still, it is used to get the id of the process. Can we do that with `concurrent`?

**Review by DanielNoord — CHANGES_REQUESTED** [review]
I pushed two changes to docstrings.

Two last (for real this time) questions.

Sorry for taking so long. Parallelism is hard and I don't want to leave any stuff from the old implementation behind after merging this.

**Review by Pierre-Sassoulas — DISMISSED** [review]
LGTM. (I was worried that the backport would conflict with #7814 but we did not merge it)

**Review by Pierre-Sassoulas — APPROVED** [review]

**Review by DanielNoord — APPROVED** [review]

**Comment by DanielNoord** [comment]
Thanks for all the work here @daniel-wer

**Pull request merged**
